### ISSUE TITLE
Change CMR endpoint for Earthdata Search Client (EDSC) extension

### ISSUE THREAD

**Issue by abarciauskas-bgse** [MAAP-Project/maap-ade-edscproxy#1]
We are in process of migrating away from MAAP having it's own CMR, so would like to change the endpoint behind the EDSC extension from search.maap-project.org to search.earthdata.nasa.gov.

@mayadebellis @bsatoriu can you advise on how to do this?

@gchang @wildintellect any concerns about moving forward with this change? I'm not proposing to deprecate search.maap-project.org just yet.

**Comment by wildintellect** [comment]
@abarciauskas-bgse Only the confusion for users. Should we have a pop up notification that this only searches NASA and not MAAP holdings? Will the code copy integration still work for notebooks?

**Comment by abarciauskas-bgse** [comment]
Definitely would look to @bsatoriu @mayadebellis to help me understand if the code copy integration still work for notebooks

Regarding confusion for users - I like the idea of a pop up but not sure how much work it would be. I was hoping if we update the documentation and send an email notification about the changes this would be just as good or better than a popup.

**Comment by wildintellect** [comment]
No one is going to remember that MAAP data is not included.

**Comment by bsatoriu** [comment]
I don't believe there is a reference to search.maap-project.org in this extension. It should be just the EDSC proxy service that we would need to point to `search.earthdata.nasa.gov`. However, we will need to test the EDSC client interface for compatibility with our code copy feature. This extension relies on a particular EDSC query string pattern for copying query context into Jupyter.

FYI Maya left JPL a couple years ago. Tagging @marjo-luc and @grallewellyn who are developing our Jupyter extensions.

**Comment by abarciauskas-bgse** [comment]
Thanks @bsatoriu - I don't know much about the EDSC proxy service but found https://github.com/MAAP-Project/maap-ade-edscproxy, I will transfer this issue there.

**Comment by grallewellyn** [comment]
I was working on the EDSC jupyter extension and running into issues (documented on a [Zenhub issue](https://app.zenhub.com/workspaces/maap-platform-5ee7f0d1f440de0024cd359b/issues/gh/maap-project/zenhub/610) but I don't believe everyone has access). It is on my to do list to take another look at this now that some of the endpoints are up and running again. The other issue I was having running edsc locally was: a bug where if I open EarthData search, and then switch to another tab on jupyter lab, I am unable to click anything (including anything in my jupyter notebook and the bar of options at the top like File, Edit, etc.). 

Also, trying to launch EarthData search client is working for me in ops ADE, but not DIT. Is anyone else experiencing this behavior? I don't think there is a ticket for this 
The behavior that I am seeing is in one workspace just nothing loaded for EarthData search client and then I switch to another workspace where I got a 404 error. Could this be a result of trying to change the CMR endpoint?

**Comment by abarciauskas-bgse** [comment]
Thanks for looking at this @grallewellyn 

I can't see the Zenhub issue, s a brief description of the EDSC jupyter extension issues here would be helpful.

I experienced the same issue as you when trying to launch EarthData search client using the Data Search dropdown option "Open EarthData Search" (works in ADE Ops but DIT I get a 404). My only guess as to why this is happening is because https://cmr.dit.maap-project.org/search/collections is returning an internal server error and I'm assuming https://search.dit.maap-project.org/search (which is also returning an error) relies on cmr.dit. _However_ I don't know how to tell which CMR endpoint the DIT environment is relying on. 

In response to the question

>Could this be a result of trying to change the CMR endpoint?

We haven't done anything yet to try and change the CMR endpoint, as far as I am aware in any existing services. The plan would be to change all the things that rely on cmr.maap-project.org to rely on cmr.earthdata.nasa.gov proactively before we deprecate cmr.maap-project.org. For NASA MAAP, those things would be:

1. The EDSC extension, which this proxy relies on, should use search.earthdata.nasa.gov
2. The default CMR_URL in maap-api-nasa (https://github.com/MAAP-Project/maap-api-nasa/blob/master/api/settings.py#L26) should be cmr.earthdata.nasa.gov. I believe maap-api-nasa what maap-py relies on as it's default `cmr_host`

Let me know if there are other places you can think of which depend on cmr.maap-project.org or search.maap-project.org.

**Comment by grallewellyn** [comment]
These links should work for Zenhub issues related to EDSC: 
https://github.com/MAAP-Project/ZenHub/issues/610 Note that this is related to developing for https://github.com/MAAP-Project/edsc-jupyter-extension. It is possible the issue about the /search links EDSC is using could be related to the 404 in DIT. I also do not know which CMR endpoint the DIT environment is using  

https://github.com/MAAP-Project/ZenHub/issues/523 Note this is related to an ADE workspace in UAT environment 

And yes, from the wording of your previous post I didn't think you made any changes, it was just the first time I noticed 'Open EarthData Search" not working in the DIT environment (not sure how long it has been like that...)

And we have been having problems with the EDSC jupyter extension since I started in September, so I am not sure if the best course of action would be to resolve those issues and then make this change, or just start using those URLs you are suggesting and try to fix the edsc problems from there. 

There are no other places I can think of that depend on cmr.maap-project.org or search.maap-project.org, but I have only worked on edsc, ipycmc, and maap help jupyter extensions.

**Comment by abarciauskas-bgse** [comment]
I would definitely suggest and request we start using search.earthdata.nasa.gov to fix these widgets. We are planning to deprecate cmr.maap-project.org (and corresponding, search.maap-project.org) by May 1. Depending on the priority of fixing these widgets, I might even suggest to remove them or have some helpful messaging that they are not expected to be working yet, until we integrate the new search.earthdata.nasa.gov and can assert that it is working.

**Comment by grallewellyn** [comment]
@marjo-luc Would you be okay with this approach?

In issue #610 I made this comment "I also tried hard coding in the actual NASA EDS site so I can making the features for the correct behavior, but this didn’t work because I get this error: The path: /lab/workspaces/edsc/proxy/https://search.earthdata.nasa.gov/search was not found" 
I will hopefully have some time tomorrow or Monday to start looking at EDSC jupyter extension, but I can start with trying to get search.earthdata.nasa.gov to work

**Comment by grallewellyn** [comment]
Just to update, I started looking into using https://search.earthdata.nasa.gov/search for EDSC, but I am running into cors policy errors and trying it with a proxy wasn't working for me. I brought this up in the MAAP meeting 4/12, and George said using https://search.earthdata.nasa.gov/search is a more complex issue and we need to discuss further. It might be something that happens after the F2F unless anyone has objections that it needs to be done sooner

**Comment by abarciauskas-bgse** [comment]
Thanks for looking into this @grallewellyn that's too bad it's a complex issue. We will need to deprecate search.maap-project.org as of May 1 so my suggestion at this point would be to remove that extension from the jupyter lab image or add an alert that the service is undergoing maintenance. Are either of those options possible by May 1?

**Comment by grallewellyn** [comment]
@marjo-luc already has a [jupyter lab image](https://github.com/MAAP-Project/maap-workspaces/blob/feature/update-versions/jupyterlab3/docker/Dockerfile) with EDSC removed, so we can do that approach. I believe that will be pushed to `develop` by the F2F
@abarciauskas-bgse

**Comment by abarciauskas-bgse** [comment]
awesome! Thanks @grallewellyn

**Comment by grallewellyn** [comment]
George said that: 
"We can't embed search.earthdata in an iframe, which is what the [EarthData Search] tab in Jupyter is. They have HTTP headers that tell your browser not to allow it. Thus, the solution is to proxy through our host, which strips out the HTTP headers before sending it to your browser. The change is done on the web server for ade.dit."
Now edsc jupyter extension is working using https://ade.dit.maap-project.org:30052/search. Note that when I first started working on edsc extension, https://ade.dit.maap-project.org:30052/search was failing to load with an internal server error, and I hope that issue doesn't continue to happen. 
Is it okay to be using https://ade.dit.maap-project.org:30052/search to access EDSC or is that what you were trying to move away from? 

Also, the three dropdowns ("Paste Collection Search Query", "Paste Granule Search Query", and "Paste Granule Search Results") aren't working because they call the [.searchCollection function in maap-py](https://github.com/MAAP-Project/maap-py/blob/master/maap/maap.py#L142) which creates requests like https://api.maap-project.org/api/cmr/collections?concept_id%5B%5D=C1214470488-ASF&concept_id%5B%5D=C2021957657-LPCLOUD&page_num=1&page_size=20 or https://api.maap-project.org/api/cmr/granules?concept_id%5B%5D=C1214470533-ASF&concept_id%5B%5D=C2021957657-LPCLOUD&concept_id%5B%5D=C1214470488-ASF&page_num=1&page_size=20 which returns a result like `"<?xml version=\"1.0\" encoding=\"UTF-8\"?><results><hits>0</hits><took>7</took><references></references></results>"` and maap-py code is expecting the collections/ granules to be inside the <results> tag. 

Looking at the [MAAP API](https://api.maap-project.org/api/), the request should be structured like `https://api.dit.maap-project.org/api/cmr/collections?concept_id[]=C1200015068-NASA_MAAP&concept_id[]=C1200090707-NASA_MAAP`, but if I replace those concept IDs with concept IDs from EarthData search, my requests like `https://api.dit.maap-project.org/api/cmr/collections?concept_id[]=C1214470488-ASF&concept_id[]=C2021957657-LPCLOUD` are still giving no results. I have tried this with many different concept IDs 

If I were to use https://cmr.earthdata.nasa.gov/search/collections instead of https://api.dit.maap-project.org/api/cmr/collections (which is maybe what @abarciauskas-bgse was suggesting before), then I am able to paste results. 
I was able to do this by making the following changes to maap-py:
1. In the MAAP class (https://github.com/MAAP-Project/maap-py/blob/master/maap/maap.py#L53), change `self._SEARCH_COLLECTION_URL = self._get_api_endpoint("search_collection_url")` to `self._SEARCH_COLLECTION_URL = 'https://cmr.earthdata.nasa.gov/search/collections'` and change `self._SEARCH_GRANULE_URL = self._get_api_endpoint("search_granule_url")` to `self._SEARCH_GRANULE_URL = https://cmr.earthdata.nasa.gov/search/granules`
2. In the CMR class (https://github.com/MAAP-Project/maap-py/blob/master/maap/utils/CMR.py#L59), change `cmr_output = response.text[1:-2].replace("\\", "")` to `cmr_output = response.text` (I am not sure what was the goal of this line, but it was causing errors in the response for me)
So now, my requests would look like this: `https://cmr.earthdata.nasa.gov/search/collections?concept_id[]=C1214470488-ASF&concept_id[]=C2021957657-LPCLOUD`
        
Is this an okay way to solve this request problem?

**Comment by gchang** [comment]
Probably shouldn't hard-code specific URLs in the code and it could mean something's up with the API in how it parses the response from CMR. Ideas @bsatoriu ?

**Comment by grallewellyn** [comment]
I was showing that as a temporary fix to demonstrate how the code would have to be changed. Although, the approach would need to be somewhat hard coded because the current self._SEARCH_COLLECTION_URL is https://api.maap-project.org/api/cmr/collections which is formed by combining [this](https://github.com/MAAP-Project/maap-py/blob/master/maap.cfg#L6) and [this](https://github.com/MAAP-Project/maap-py/blob/master/maap.cfg#L12) from maap.cfg. The second one would be easy to change since it is specific, but the first one would require a special case when to use a certain host (ie cmr.earthdata.nasa.gov)

**Comment by bsatoriu** [comment]
@grallewellyn my proposal is to change the CMR url in the MAAP API service: https://github.com/MAAP-Project/maap-api-nasa/blob/develop/api/settings.py#L18. This would fix the core server error and leave the client maap.cfg file unchanged.

**Comment by grallewellyn** [comment]
@bsatoriu So I should change line `CMR_URL = 'https://cmr.maap-project.org'` to `CMR_URL = 'https://cmr.earthdata.nasa.gov/`?
How can I test this to make sure it works for edsc? Would running the maap-api-nasa README help me test edsc?

**Comment by bsatoriu** [comment]
Yes, you can test locally using the README instructions. Let me know if you need any help getting it running.

**Comment by grallewellyn** [comment]
I was able to change the CMR_URL to https://cmr.earthdata.nasa.gov and successfully run `docker-compose up` for maap-api-nasa (but I did have to change the `docker-compose.yml` ports line to 8080:5000 because I was getting an error about port 5000 still being in use despite netstat not having any results).

I am able to access http://0.0.0.0:8080/api/ which is my locally hosted version of the page that looks like https://api.dit.maap-project.org/api/. 
The problem pasting search results is because requests like https://api.dit.maap-project.org/api/cmr/collections?concept_id[]=C1214470488-ASF&concept_id[]=C2021957657-LPCLOUD aren't returning anything in the results tag. If I change the url to http://0.0.0.0:8080/api/cmr/collections?concept_id[]=C1214470488-ASF&concept_id[]=C2021957657-LPCLOUD, I am also not getting anything in the results tag, so I am not sure if changing the CMR_URL will solve the problem. What are your thoughts @bsatoriu?

**Comment by bsatoriu** [comment]
@grallewellyn I updated the CMR_URL variable in our GIT Gitlab CI repo and the search results seem to work for me now: [https://api.dit.maap-project.org/api/cmr/collections?concept_id[]=C1214470488-ASF&concept_id[]=C2021957657-LPCLOUD](https://api.dit.maap-project.org/api/cmr/collections?concept_id%5B%5D=C1214470488-ASF&concept_id%5B%5D=C2021957657-LPCLOUD)

I'm not sure why it's not working locally for you. Were you able to check the values in the local debugger?

**Comment by grallewellyn** [comment]
@bsatoriu How do I check the values in the local debugger? 
I only see some warnings in my terminal where I ran `docker-compose up` and I don't see anything in the docker logs. 

That link also works for me, but running `maap.searchCollection(concept_id="C1214470488-ASF|C2021957657-LPCLOUD", limit=1)` in a jupyter notebook still gives no results. When I copy in my own MAAP and CMR instance into a notebook, it seems that is because the url before adding params in [get_search_results](https://github.com/MAAP-Project/maap-py/blob/1bfa99d4aa3493a5000de05829b77096843d3967/maap/utils/CMR.py#L20) is https://api.maap-project.org/api/cmr/collections not https://api.dit.maap-project.org/api/cmr/collections (ie doesn't have dit, maybe it is ops?)

A request like https://api.maap-project.org/api/cmr/collections?concept_id%5B%5D=C1214470488-ASF&concept_id%5B%5D=C2021957657-LPCLOUD gives an empty results tag. Do you have any recommendations for either 
1) Making https://api.maap-project.org/api/ behave the same as the one with dit and have results in the results tag
2) Altering maap-py but not the maap.cfg so that https://api.dit.maap-project.org/api/cmr/collections is used instead of the one without dit

**Comment by grallewellyn** [comment]
@bsatoriu Just following up again on the above 
When I posted that https://api.maap-project.org/api/cmr/collections?concept_id%5B%5D=C1214470488-ASF&concept_id%5B%5D=C2021957657-LPCLOUD was giving an empty results tag, but now it is giving `{"errors": ["An Internal Error has occurred."]}`

@abarciauskas-bgse Is it okay if we continue using https://api.maap-project.org/api/cmr/collections or https://api.dit.maap-project.org/api/cmr/collections instead of changing to https://cmr.earthdata.nasa.gov/search/collections?

**Comment by wildintellect** [comment]
@grallewellyn yes our MAAP CMR cluster went down this week, and we don't know why (was scheduled to be turned off next week anyways. If api.maap-project.org is a proxy wrapper around cmr.earthdata, that should be fine for now.

**Comment by grallewellyn** [comment]
Before a request like https://api.maap-project.org/api/cmr/collections?concept_id%5B%5D=C1214470488-ASF&concept_id%5B%5D=C2021957657-LPCLOUD was giving an empty results tag, but now it is working! I am able to successfully run all the dropdowns for EDSC jupyter extension which means I can run like `maap.searchGranule(concept_id="C1214470488-ASF", limit=1000)` in the notebook. 
I just need to make a couple more tweaks to improve the extension and then I can do some testing before deploying

**Comment by grallewellyn** [comment]
Also, confirmed in the meeting that api.maap-project.org is a proxy wrapper around cmr.earthdata, and we should be fine for now

**Comment by grallewellyn** [comment]
Right now the paste Granule Search for EDSC jupyter extension isn't working. The problem seems to be that https://api.dit.maap-project.org/api/cmr/granules?concept_id=C1214470488-ASF&page_num=1&page_size=20 is giving an access denied error. Either:
1. The permission denied error is fixed
2. Maap-py always uses https://api.maap-project.org/api/cmr/granules?concept_id=C1214470488-ASF&page_num=50&page_size=20 instead because this is behaving correctly. I don't even know if it is possible to make this change, but it would maybe involve changing the environments.json 

@bsatoriu Do you know why this permission denied error is happening?
